crates/wasmtime/src/runtime/component/func/host.rs

```diff
@@ -12,11 +15,75 @@ use alloc::sync::Arc;
 use core::any::Any;
 use core::mem::{self, MaybeUninit};
 use core::ptr::NonNull;
+#[cfg(feature = "rr-type-validation")]
+use wasmtime_environ::component::TypeTuple;
 use wasmtime_environ::component::{
     CanonicalAbiInfo, ComponentTypes, InterfaceType, MAX_FLAT_PARAMS, MAX_FLAT_RESULTS,
     StringEncoding, TypeFuncIndex,
 };
 
+/// Record/replay stubs for host function entry events
+macro_rules! rr_host_func_entry_event {
+    { $args:expr, $param_types:expr => $store:expr } => {
+        #[cfg(any(feature = "rr-type-validation", feature = "rr-args-validation"))]
+        {
+            use crate::config::ReplayMetadata;
+            use crate::runtime::rr::events::component_wasm::HostFuncEntryEvent;
+            $store.record_event_if(
+                |r| r.add_validation,
+                |_| {
+                    HostFuncEntryEvent::new(
+                        $args,
+                        #[cfg(feature = "rr-type-validation")]
+                        Some($param_types),
+                    )
+                },
+            );
+            $store.next_replay_event_if(
+                |_, r| r.add_validation,
+                |_event: HostFuncEntryEvent, _r: &ReplayMetadata| {
+                    #[cfg(feature = "rr-type-validation")]
```

Comment:
Rather than conditionalizing individual statements, generally what we try to do for this sort of thing is conditionalize on function definitions, and write alternate (often empty) ones -- this keeps the flow a bit cleaner inside bodies. So something like:

```
#[cfg(feature = "rr-type-validation")]
fn validate(...) { /* actually do it */ }

#[cfg(not(feature = "rr-type-validation"))]
fn validate(...) { /* nothing */ }
```

on `event`'s impl, then always write `if r.validate { event.validate(...); }` here. This also avoids the awkward unused-variable warnings you're working around with the underscores here.